cmd/kube-proxy/app/server_others.go

```diff
@@ -145,6 +152,28 @@ func newProxyServer(
 		return nil, fmt.Errorf("cannot determine detect-local-mode: %v", err)
 	}
 
+	var nodeInfo *v1.Node
+	if detectLocalMode == proxyconfigapi.LocalModeNodeCIDR {
+		klog.Infof("Watching for node %s, awaiting podCIDR allocation", hostname)
+
+		// since allocators can assign the podCIDR after the node registers, we do a watch here to wait
+		// for podCIDR to be assigned, instead of assuming that the Get() on startup will have it.
+		ctxt, cancelContextFn := context.WithTimeout(context.TODO(), timeoutForNodePodCIDR)
+		w, err := client.CoreV1().Nodes().Watch(ctxt, metav1.ListOptions{FieldSelector: fmt.Sprintf("metadata.name=%s", hostname)})
+		if err != nil {
+			klog.V(0).Infof("Error setting watch on node: %v", err)
+			return nil, fmt.Errorf("unable to set watch on node to configure detect-local-mode %v", proxyconfigapi.LocalModeNodeCIDR)
+		}
+		nodeChan := make(chan *v1.Node)
+		go watchForPodCIDR(ctxt, w, nodeChan)
```

Comment:
Do we really need a goroutine for this?  Why not just a function that returns a value?

Reply:
since we need a context anyway to make the call to watch, it simplified the timeout handing, and also handles the case where watch itself hangs graciously while enforcing the timeout.

Reply:
removed the goroutine.